Disconnect on a malformed version message, and reject duplicate verack messages

daira opened the pull request:
This is loosely based on and supercedes #6780 by @AndreaLanfranchi, but takes a different approach. It makes the following changes:

* Deserialization of fields of a "version" message now uses types of platform-independent size.
* If deserialization of a "version" message fails due to a `std::ios_base::failure`, the "version" message will be rejected and the peer will be marked to disconnect (`pfrom->fDisconnect = true`).
* A peer's `nVersion` and `nServices` fields are no longer set (i.e. they will be left at 0) until after the mandatory fields of a "version" message have been successfully parsed and validated. Therefore, the `nVersion` field will never be set to a non-zero value below `MIN_PEER_PROTO_VERSION` (or `MIN_TESTNET_PEER_PROTO_VERSION` for Testnet). This makes the peer version handling simpler to reason about.
* `pfrom->fSuccessfullyConnected` is not set until after `pfrom->nTimeOffset` has been initialized.
* If a duplicate "verack" message is received, it will be rejected in the same way as a duplicate "version" message.

daira committed:
Remove code that is definitely dead, because pfrom->nVersion must be
>= MIN_PEER_PROTO_VERSION (170100) at this point.

Signed-off-by: Daira Emma Hopwood <daira@jacaranda.org>

daira committed:
Fix some issues with handling of version messages:
* fields were deserialized into variables of platform-dependent length;
* pfrom->nVersion was set even if it does not meet the minimum version
  requirements;
* make sure that pfrom->nTimeOffset is set before
  pfrom->fSuccessfullyConnected.

At this point it is still possible for parsing to fail due to a
`std::ios_base::failure`, which will not cause a disconnect.

Co-authored-by: Andrea Lanfranchi <andrea.lanfranchi@gmail.com>
Signed-off-by: Daira Emma Hopwood <daira@jacaranda.org>

daira committed:
Treat a malformed version message as cause for disconnection.

Co-authored-by: Andrea Lanfranchi <andrea.lanfranchi@gmail.com>
Signed-off-by: Daira Emma Hopwood <daira@jacaranda.org>

daira committed:
Reject duplicate verack.

Co-authored-by: Andrea Lanfranchi <andrea.lanfranchi@gmail.com>
Signed-off-by: Daira Emma Hopwood <daira@jacaranda.org>

daira commented:
@AndreaLanfranchi I used "Co-authored-by: Andrea Lanfranchi <andrea.lanfranchi@gmail.com>" to attribute you as a co-author for the last three commits. However I can't sign those with your gpg key, and you have "vigilant mode" set (which warns if any commit attributed to you is not signed by your key). Would you prefer me to remove those per-commit attributions and only acknowledge you in the PR description, or is that okay as-is despite the "Unverified" warning?

nuttycom commented on src/main.cpp:
```diff
@@ -6835,37 +6836,44 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
+        pfrom->nVersion = nVersion;
+        pfrom->nServices = nServices;
+
+        if (!vRecv.empty()) {
```
Should these check the length remaining to ensure sufficient bytes remain for successful parsing? Or is the approach here to rely upon exception handling?

daira replied:
The length of the message is allowed to be either the minimum length (just the mandatory fields), or exactly one of the lengths obtained by adding each optional set of fields in turn, or anything longer than the longest such length (to allow for future expansion). That's exactly what the current code allows, and yes, we're relying on exception handling to reject anything else.

nuttycom reviewed (approved):
utACK 54bdbf5e3ba50d987d9595390ba98761cd724258

daira commented:
Since this potentially affects peer-to-peer protocol compatibility (and involves parsing code and exceptions in C++), I'd like two reviews please :-)

daira commented:
Incidentally, I also checked that `nVersion` and `nStartingHeight` were the only places in `ProcessMessage` where we deserialized into a platform-dependent type (`int` in both cases). We don't support platforms with `int` less than 32 bits, but `int` technically could be longer.

AndreaLanfranchi commented:
> @AndreaLanfranchi I used "`Co-authored-by: Andrea Lanfranchi <andrea.lanfranchi@gmail.com>`" to attribute you as a co-author for the last three commits. However I can't sign those with your gpg key, and you have "vigilant mode" set (which warns if any commit attributed to you is not signed by your key). Would you prefer me to remove those per-commit attributions and only acknowledge you in the PR description, or is that okay as-is despite the "Unverified" warning?

Hi @daira,
I have no problems with either choice. It mostly depends on repo constraints whether it allow the application of unverfieid commits. And thank you for asking.

daira commented:
@AndreaLanfranchi wrote:
> I have no problems with either choice.

In that case it can stay as-is.

str4d commented on src/main.cpp:
```diff
@@ -6975,12 +6989,18 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
 
     else if (strCommand == "verack")
     {
+        LOCK(cs_main);
+        CNodeState* state = State(pfrom->GetId());
+        if (state->fCurrentlyConnected) {
+            pfrom->PushMessage("reject", strCommand, REJECT_DUPLICATE, std::string("Duplicate verack message"));
+            Misbehaving(pfrom->GetId(), 1);
+            return false;
+        }
```
I confirmed that `State()` is used elsewhere without being guarded by `pfrom->fNetworkNode` (which makes sense as it is intended for decoupling node-relevant network state from the network locks).

Currently the new reject will never trigger when a duplicate `verack` message is received from a non-`fNetworkNode`, which AFAICT includes all inbound connections (`fNetworkNode` is only set to true in `OpenNetworkConnection` for outbound connections). This implies that if we want to reject duplicate verack messages from all non-local `CNode`s (both outbound and inbound), then we need a new boolean to track that state (because `fCurrentlyConnected` is from upstream, and so we shouldn't go changing its semantics inside the node).

AndreaLanfranchi replied:
I believe the presence, actually, of two separate booleans (`fCurrentlyConnected` from State and `fSuccessfullyConnected` from CNode member) are a bit confusing and, IMHO, misleading. Wrt specifically to the latter its meaning is ambiguous as it may appear as a redundant statement of what already is in State but at the same time it doesn't mean the handshake process has been fully completed (exchange amongst the two ends of the socket of **both** `version` and `verack`). 
I believe it would be beneficial to move the valuation of `fSuccessfullyConnected` to true only after `verack` message has been successfully parsed and processed **and** at the same time move there as well :

1. the `getaddr` request (if applicable)
2. the advertising of local address (if applicable)
3. the tagging as "good" (in addrman) of the sender's address (if applicable): actually a "good" address, IMHO, should be capable at least to complete the handshake

All of this should be helpful to solve issues #5168 and #6038 (protocol handshake should NOT be interleaved by any message other than version and verack).

AndreaLanfranchi replied:
Also worth mentioning on the matter that latest bitcoin core implementation totally disregards the 26 bytes of the "from" address in version message while acknowledging the remote peer address from its addr messages.

daira replied:
As #6038 demonstrates, we cannot drop messages that are received before we have sent "verack", because previous versions of zcashd send them and expect a response. We can stop sending them before the handshake has completed, but we must accept them when received.

daira replied:
Anyway, #5168 and #6038 are a separate issue or issues that we should not attempt to fix in this PR, which is already complicated enough to analyse. I'm with @str4d that we should not be changing the semantics of variables that are present in Bitcoin Core.

AndreaLanfranchi replied:
> As #6038 demonstrates, we cannot drop messages that are received before we have sent "verack", because previous versions of zcashd send them at that point and expect a response. We can stop sending them before we have received "verack", but we must accept them when received.

I think abstain from consuming them before the handshake has been completed in full could be a valuable intermediate solution. Anyhow, just for my knowledge and education, [latest bitcoin core's code base have moved the valuation of `fSuccessfullyConnected` flag from CNode's into the processing of `verack` message](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L3592-L3642).

In any case I'm not pushing to have more changes in this PR specifically.

str4d reviewed (commented):
Reviewed as of 54bdbf5e3ba50d987d9595390ba98761cd724258.

nuttycom commented on src/main.cpp:
```diff
@@ -6975,12 +6989,18 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
     else if (strCommand == "verack")
     {
+        LOCK(cs_main);
+        CNodeState* state = State(pfrom->GetId());
```
Why are we certain that `state` will not be a null pointer at this point? I note that elsewhere in this method, we rely upon that invariant, but it's not obvious to me why it holds.

daira replied:
The `State` function can only return null when `pnode` is not a key in `mapNodeState`. There are other cases in handling messages received from a peer `CNode *pfrom` where we assume that `State(pfrom->GetId())` will be non-null. For example, in the processing of "version":

```c++
        // Potentially mark this peer as a preferred download peer.
        {
            LOCK(cs_main);
            UpdatePreferredDownload(pfrom, State(pfrom->GetId()));
        }
```
where
```c++
void UpdatePreferredDownload(CNode* node, CNodeState* state)
{
    nPreferredDownload -= state->fPreferredDownload;
[...]
```

However, in investigating this I decided it would be a good idea to backport bitcoin/bitcoin#10176, and also to defensively add an assertion (to make the case of a null pointer here a crash, rather than undefined behaviour).

nuttycom reviewed (dismissed):
utACK 09637cd with nonblocking question.

daira committed:
net: define NodeId as an int64_t

This should make occurrences of NodeId wrapping essentially impossible for
real-world usage.

Backport of https://github.com/bitcoin/bitcoin/pull/10176

[zcashd] I have checked that zcashd has no current uses of NodeId that
depend on it being an `int`. All accesses to the global `nLastNodeId`
are under lock. `NodeId` *is* required to be a signed integral type,
because `-1` is used as a sentinel value. It is also formatted using
the `%d` tinyformat specifier, but unlike the C format specifier it is
inspired by, this correctly handles integral types of arbitrary width.

There are `NodeId` fields in `CNodeStats`, `NodeEvictionCandidate`, and
(test-only) `COrphanTx`, but those types are not serializable, and there
is no other ad-hoc serialization of `NodeId` values apart from its use
in the "id" field of the output from the `getpeerinfo` RPC. `UniValue` has
an override of `pushKV` for `int64_t`, and so that use will correctly
handle values up to the [maximum safe JSON/JavaScript integer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER),
i.e. $2^{53} - 1$. As upstream did, we argue that it is not feasible to
cause that value to be exceeded.

Co-authored-by: Cory Fields <cory-nospam-@coryfields.com>
Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

str4d reviewed (approved):
utACK abe934ce79af8c0a6ca3fb4b09b8baa958159a3d.

I re-confirmed that, per my above comment, this PR only rejects duplicate verack messages for outbound connections (because it reuses the existing `state->fCurrentlyConnected` boolean which is used for tracking outbound connection attempts to addresses in `addrman`). If duplicate verack rejection is desired for inbound connections, that will need to be implemented in a separate PR with a new boolean that isn't coupled to outbound connection attempts.

illuminated-shade reviewed (commented):
src/net.h
